Add checksum computation after fetch to multiple cluster modules

JavierJF opened the pull request:
Add the computation of the checksum of the fetched tables for the following config modules:

- mysql_servers
- global_variables
- mysql_users
- mysql_query_rules

This ensures that the fetched information is correct and prevent race conditions between the checksum fetching and the actual module information fetching.

Flagging the PR as WIP since at this stage LDAP support from LDAP plugin is pending

JavierJF committed:
Add checksum computation after fetch to multiple cluster modules

The following cluster modules now compute their expected checksums after fetch:
  - mysql_query_rules
  - mysql_users
  - mysql_servers
  - mysql_global_variables

JavierJF committed:
Fix potential race conditions while generating 'gtid_executed_tables'

JavierJF committed:
Fix invalid memory accesses in Aurora monitoring

Invalid memory accesses could take place if no servers can be pinged.

JavierJF committed:
Add utility function to convert 'MYSQL_RES' into 'SQLite3_result'

JavierJF committed:
Improve 'test_cluster_sync' including 'mysql_servers'

JavierJF commented:
Retest this please

JavierJF committed:
Prevent Cluster sync between nodes with/without 'ldap_auth' plugin loaded

JavierJF committed:
Add LDAP support for 'mysql_users' Cluster checksum computation

JavierJF committed:
Fix crash during startup due to 'GloClickHouseServer' not being yet initialized

An invalid memory access can take place if 'GenericRefreshStatistics' is
called during startup when 'GloClickHouseServer' isn't yet initialized.

JavierJF committed:
Removed invalid load from disk of 'mysql_ldap_mapping' in 'ProxySQL_Admin::init'

JavierJF committed:
Fix test 'test_cluster1' synchronization for 'mysql_servers'

renecannao commented on lib/MySQL_HostGroups_Manager.cpp:
```diff
@@ -2124,6 +2124,9 @@ void MySQL_HostGroups_Manager::generate_mysql_gtid_executed_tables() {
 
+	// NOTE: We are required to lock while iterating over 'MyHostGroups'. Otherwise race conditions could take place,
+	// e.g. servers could be purged by 'purge_mysql_servers_table' and invalid memory be accessed.
+	wrlock();
```
Good catch!

renecannao commented on lib/ProxySQL_Admin.cpp:
```diff
@@ -11380,8 +11397,8 @@ void ProxySQL_Admin::save_mysql_servers_runtime_to_database(bool _runtime) {
 				rc=(*proxy_sqlite3_bind_int64)(statement32, (idx*12)+4, atoi(r1->fields[3])); ASSERT_SQLITE_OK(rc, admindb);
-				rc=(*proxy_sqlite3_bind_text)(statement32, (idx*12)+5, ( _runtime ? r1->fields[5] : ( strcmp(r1->fields[5],"SHUNNED")==0 ? "ONLINE" : r1->fields[5] ) ), -1, SQLITE_TRANSIENT); ASSERT_SQLITE_OK(rc, admindb);
-				rc=(*proxy_sqlite3_bind_int64)(statement32, (idx*12)+6, atoi(r1->fields[4])); ASSERT_SQLITE_OK(rc, admindb);
+				rc=(*proxy_sqlite3_bind_text)(statement32, (idx*12)+5, ( _runtime ? r1->fields[4] : ( strcmp(r1->fields[4],"SHUNNED")==0 ? "ONLINE" : r1->fields[4] ) ), -1, SQLITE_TRANSIENT); ASSERT_SQLITE_OK(rc, admindb);
```
Good catch!

renecannao commented on lib/ProxySQL_Admin.cpp:
```diff
@@ -4476,18 +4476,34 @@ void admin_session_handler(MySQL_Session *sess, void *_pa, PtrSize_t *pkt) {
+		if (GloMyLdapAuth == nullptr) {
+			sprintf(query, q, PROXYSQL_VERSION);
+		} else {
+			sprintf(query, q, PROXYSQL_VERSION" Enterprise");
```
I would suggest a `-` (dash) and not a space.

renecannao commented on lib/ProxySQL_Admin.cpp:
```diff
@@ -4476,18 +4476,34 @@ void admin_session_handler(MySQL_Session *sess, void *_pa, PtrSize_t *pkt) {
+		if (GloMyLdapAuth == nullptr) {
+			sprintf(query, q, PROXYSQL_VERSION);
+		} else {
+			sprintf(query, q, PROXYSQL_VERSION" Enterprise");
```
I would suggest a `-` (dash) and not a space.

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -112,8 +114,9 @@ void * ProxySQL_Cluster_Monitor_thread(void *args) {
 						if (row[0]) {
-							if (strcmp(row[0], PROXYSQL_VERSION)==0) {
-								proxy_info("Cluster: clustering with peer %s:%d . Remote version: %s . Self version: %s\n", node->hostname, node->port, row[0], PROXYSQL_VERSION);
+							const char* PROXYSQL_VERSION_ = GloMyLdapAuth == nullptr ? PROXYSQL_VERSION : PROXYSQL_VERSION" Enterprise";
```
I would suggest a `-` (dash) and not a space.

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -863,6 +921,12 @@ void ProxySQL_Cluster::pull_mysql_query_rules_from_peer() {
+
+						const uint64_t query_rules_hash = mysql_raw_checksum(result1) + mysql_raw_checksum(result2);
+						const string computed_checksum { get_checksum_from_hash(query_rules_hash) };
+
```
We can add a new `proxy_info()` to display the computed checksum.
```
proxy_info("Cluster: Computed checksum for MySQL Query Rules from peer %s:%d : %s\n", hostname, port, computed_checksum.c_str() );

JavierJF replied:
Done! Also added info when any of the checksum are computed via "LOAD *" commands, to ensure that the locally computed checksum is also present in the log.

renecannao commented on lib/gen_utils.cpp:
```diff
@@ -220,3 +226,30 @@ bool Proxy_file_regular(const char *path) {
 	return false;
 }
 
+std::unique_ptr<SQLite3_result> get_SQLite3_resulset(MYSQL_RES* resultset) {
```
If this function is not used, we should disable (for code coverage)

renecannao replied:
Actually, I may be using in another PR. Please hold

JavierJF committed:
Merge branch 'v2.x' of github.com:sysown/proxysql into v2.x-cluster_cmp_checksums

JavierJF committed:
Perform atomic update of Cluster fetched module checksum epochs

JavierJF committed:
Revert "Fix test 'test_cluster1' synchronization for 'mysql_servers'"

This reverts commit 13a7d7fab0b5addcc4fe2d32515142412daccc6e.

JavierJF committed:
Cluster: Speedup of processing of MySQL Users and MySQL Servers

- Cluster now syncs the server tables via 'incoming_*' tables generated
  during 'load_mysql_servers_to_runtime'.
- Cluster now syncs 'mysql_users' table via a resultset generated via
  '__refresh_users'.

JavierJF committed:
Fix 'test_cluster1-t' checksum fetching logic

JavierJF committed:
Fix minor typos in metrics descriptions

JavierJF committed:
Fix memory leak for 'Scheduler_Row::filename' field

JavierJF committed:
Fix memory leaks for 'username' and 'password' credentials in cluster 'pull' actions

JavierJF committed:
Fix memory leaks in 'Query_Processor' for Query Rules 'attributes' and 'comment' fields

JavierJF committed:
Change servers entries order in 'test_cluster_sync-t' to test checksum computation

JavierJF committed:
Replace space in ProxySQL version Enterprise identifier in favor of dash

JavierJF committed:
Add 'computed checksum' info messages before performing pulling actions in Cluster

JavierJF committed:
Fix compilation adding constructor for 'incoming_servers_t'

A constructor is required to overcome C++11 limitation of type
no longer being an aggregate due to default member initialization.

JavierJF committed:
Replace resultset for Cluster request fetching 'runtime_mysql_servers'

Query 'CLUSTER_QUERY_MYSQL_SERVERS' is now responded by 'runtime_mysql_servers'
as it previously was. Documentation reasoning about the change has been added
to 'MySQL_HostGroups_Manager::runtime_mysql_servers'.

JavierJF committed:
Filter 'OFFLINE_HARD' servers for stored 'runtime_mysql_servers' resultset

JavierJF committed:
Change wording in doc for 'MySQL_HostGroups_Manager::runtime_mysql_servers'

renecannao commented:
retest this please

JavierJF committed:
Fix typos in some cluster queries definitions doc
